firedrake/meshadapt.py

```diff
@@ -0,0 +1,500 @@
+import abc
+from collections.abc import Iterable
+import firedrake
+import firedrake.function as ffunc
+import firedrake.functionspace as ffs
+import firedrake.mesh as fmesh
+from firedrake.petsc import PETSc, OptionsManager
+import firedrake.utils as futils
+import numpy as np
+import ufl
+
+__all__ = ["RiemannianMetric", "MetricBasedAdaptor", "adapt"]
+
+
+class RiemannianMetric(ffunc.Function):
```

Comment:
Can you explain why you are inheriting from a `Function` here? It looks like you are doing a lot of complicated logic here like modifying the underlying plex that makes me think that a subclass isn't the right approach. Could this class not just have a `function` property instead? (i.e. composition instead of inheritance).

Reply:
This was how it was initially done in the previous PR, but as we discussed it we realised there are quite a few `Function` methods that it is useful to carry over, in particular `assign`, `interpolate`, `project`. Also, a Riemannian metric is just a special case of a tensor field that is SPD at each point. From that perspective, it makes sense for it to be a subclass. Open to other thoughts and suggestions, though!

Reply:
Symmetric positive-definite, i.e. symmetric matrix with all of its eigenvalues being positive.